### DISCUSSION TITLE
Cleaning Program Names from special characters at StateData step.

### DISCUSSION THREAD

**Discussion by damiensellier** [RomanKubiak/ctrlr#286] [Q&A] [answered]
Hi everybody, 

I am finishing my project and at the last moment I discovered a bug.
User Bank Program Names are stored in a table called UserPrgrmNameTable.
When recalling the User Bank Program Names, if I have other characters than letters or spaces, my string stored in the value tree is shifting the Names strangely.

I don't know where to find the Yamaha characters table but for exemple : 
DE 4C 4F 57 DE 53 4E 41 52 45 20 20 20 20 20 20
DE is quotation mark for yamaha 
it should be "LOW"SNARE

While converted to string via LUA it gives something like фLOWфSNARE  with cyrilic characters or something non occidental.
If i save the program in a sysex file this is my output name : 46 - ÃÂÃÂLOWÃÂÃÂS
Which is totally messed up.

when stored in the valueTree as a string it breaks the tree because the quotation mark is not properly recognised and mistaken for a dot or something else.

How can I only keep the letters and skip the special yamaha characters to avoid this issue?

Here is my scripts

When I convert the hex values to Program names strings , i use the usual suspect : 

```
function RetrieveUserProgramNameDecode(MbUserPrgrmNameRaw)
	PSTchar1 = MbUserPrgrmNameRaw:getByte(0)
	PSTchar2 = MbUserPrgrmNameRaw:getByte(1)
	PSTchar3 = MbUserPrgrmNameRaw:getByte(2)
	PSTchar4 = MbUserPrgrmNameRaw:getByte(3)
	PSTchar5 = MbUserPrgrmNameRaw:getByte(4)
	PSTchar6 = MbUserPrgrmNameRaw:getByte(5)
	PSTchar7 = MbUserPrgrmNameRaw:getByte(6)
	PSTchar8 = MbUserPrgrmNameRaw:getByte(7)
	PSTchar9 = MbUserPrgrmNameRaw:getByte(8)
	PSTchar10 = MbUserPrgrmNameRaw:getByte(9)
	PSTchar11 = MbUserPrgrmNameRaw:getByte(10)
	PSTchar12 = MbUserPrgrmNameRaw:getByte(11)
	PSTchar13 = MbUserPrgrmNameRaw:getByte(12)
	PSTchar14 = MbUserPrgrmNameRaw:getByte(13)
	PSTchar15 = MbUserPrgrmNameRaw:getByte(14)
	PSTchar16 = MbUserPrgrmNameRaw:getByte(15)

	PSTchar1ascii = string.char(PSTchar1)
	PSTchar2ascii = string.char(PSTchar2)
	PSTchar3ascii = string.char(PSTchar3)
	PSTchar4ascii = string.char(PSTchar4)
	PSTchar5ascii = string.char(PSTchar5)
	PSTchar6ascii = string.char(PSTchar6)
	PSTchar7ascii = string.char(PSTchar7)
	PSTchar8ascii = string.char(PSTchar8)
	PSTchar9ascii = string.char(PSTchar9)
	PSTchar10ascii = string.char(PSTchar10)
	PSTchar11ascii = string.char(PSTchar11)
	PSTchar12ascii = string.char(PSTchar12)
	PSTchar13ascii = string.char(PSTchar13)
	PSTchar14ascii = string.char(PSTchar14)
	PSTchar15ascii = string.char(PSTchar15)
	PSTchar16ascii = string.char(PSTchar16)

	UserPrgrmName = string.format("%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s%s", PSTchar1ascii, PSTchar2ascii, PSTchar3ascii, PSTchar4ascii, PSTchar5ascii, PSTchar6ascii, PSTchar7ascii, PSTchar8ascii, PSTchar9ascii, PSTchar10ascii, PSTchar11ascii, PSTchar12ascii, PSTchar13ascii, PSTchar14ascii, PSTchar15ascii, PSTchar16ascii)
end
```

Then The names are stored in a table "UserPrgrmNameTable" while in a loop with a table insert : 

```
..
table.insert(UserPrgrmNameTable,UserPrgrmName)
..
```

When I quit the app

```
SaveStateData = function(--[[ ValueTree --]]stateData)
	if UserPrgrmNameTable ~= nil then
		sUserPrgrmName = table.concat(UserPrgrmNameTable)
	end

	stateData:setProperty("UserPrgrmName", sUserPrgrmName, nil)

end
```

When I launch the app
 
```
ReadStateData = function(--[[ ValueTree --]] stateData)

	if stateData:hasProperty("UserPrgrmName") == true then 
		sUserPrgrmName = tostring(stateData:getProperty("UserPrgrmName"))		
		UserPrgrmNameTable = {}

		for i=0,99 do
			UserPrgrmNameValue = string.sub((sUserPrgrmName), (i*16)+1 , (i*16)+16 ) -- split the valuetree string to 16 char long Prgrm Name
			table.insert(UserPrgrmNameTable,UserPrgrmNameValue)
    		end
	end -- if
end-- f
```

I don't know at which stage I should introduce the Program Name cleaning process. 
Is it beforehand when I convert hex values to the  strings stored in the UserPrgrmNameTable?
Is it with SaveStateData when quiting the app?
Is it at the ReadStateData stage on launch?

Thanks a lot for your help, have a good week.

Damien

This is what I have when I populate the table with names from sysex dumps of prgrm 0 to 99:
![Preset names stored from sysex](https://user-images.githubusercontent.com/55164385/117736031-b5fe9980-b1c4-11eb-9f60-b31697eb0043.jpg)

This is what i get when I restore the names from the valuetree string with ReadStateData  when I launch the panel again : 
![Preset names restored from ValueTree](https://user-images.githubusercontent.com/55164385/117736089-cf9fe100-b1c4-11eb-9862-b5a5755b4fde.jpg)

**Comment by dnaldoog** [comment]
<code>if stateData:hasProperty("UserPrgrmName") == true then</code>

This is new to me! Thank you! All along I have been doing :
<pre>
<code>
local sUserPrgrmName = stateData:getProperty("UserPrgrmName")
if sUserPrgrmName ~= nil then
-- do stuff
end
</code>
</pre>

----
DE 4C 4F 57 DE 53 4E 41 52 45 20 20 20 20 20 20
DE does print ÞLOWÞSNARE   

so it looks like you would need some kind of lookup table that converts it properly for the panel and re converts it back for the device:

Converting for the panel would go something like:
<pre>
<code>
function  forPanel(n) -- pass in a string
n=string.upper(n)
local t={
["DE"]='"',
["20"]=" "
["4C"]="L"
}
return t[n]
end
</code>
</pre>
----
To go the other way:

<pre>
<code>
function  forSynth(n) -- pass in a string
n=string.upper(n)
t={
  ['"']="DE" -- or 0xDE
  L="4C" --or 0x4c
}
return t[n]
end
</code>
</pre>

print("for synth",forSynth("L")) = 4C
print("for synth",forSynth('"')) = DE

**Reply by damiensellier** [reply]
Thanks for your help John. The lookup table is a nice idea. I finally got the yamaha character table in the manual but the position has nothing to do with the hex value of the character so I had to forget that way. 

The problem occures when the Program Names are restored from the valueTree so I decided to cleanup the character strings before saving on quit.

I used the string substitute for non alphanumeric characters to "space" like this with  string:gsub('%W',' ') : 

```
SaveStateData = function(--[[ ValueTree --]]stateData)
	if UserPrgrmNameTable ~= nil  then
		sUserPrgrmName = table.concat(UserPrgrmNameTable)
		sUserPrgrmName = sUserPrgrmName:gsub('%W',' ')
	end

	stateData:setProperty("UserPrgrmName", sUserPrgrmName, nil)

end
```

Et voilà, since it's just for display purpose I don't care if the quote marks are replaced with spaces for exemple. 
Now I have every name at their right position in the table. So it's a success.

Have a nice week.

Damien
![Yamaha Character table](https://user-images.githubusercontent.com/55164385/117745357-514d3a00-b1d8-11eb-9dc8-866b1f41b362.png)

**Comment by damiensellier** [comment, marked as answer]
Solution : 
To replace non alphanumeric characters by a "space" use string gsub : 
`string:gsub('%W',' ')`

Note : An upper case version of any of the classes (w for alphanumerical) represents the complement of the class. For instance, '%W' represents all non-alpha character.

**Reply by dnaldoog** [reply]
That’s a very nice solution considering there are also katakana characters etc included in the character set.
